### ISSUE TITLE
Fix LinkedTreeSet Iterator#remove()

### ISSUE THREAD

**Issue by markbigler** [bwaldvogel/mongo-java-server#109]
I stumbled across a nasty bug while running the following expression: `Expression.evaluate(json("'$size': {'$setIntersection': ['$tags', ['B']]}"), json("tags: ['A', 'B']")))`

Expected size is `1` but a size of `2` is returned.

It looks like the implementation of `de.bwaldvogel.mongo.backend.LinkedTreeSet$Iterator#remove()` is causing this issue, because it only removes elements from `orderedElements` but not from `elements`. Since `de.bwaldvogel.mongo.backend.LinkedTreeSet#size()` is based on `elements`, this leads to a wrong result.

Test cases to reproduce the issue:
```java
@Test
public void testSizeAfterRemoveAll() throws Exception {
    Set<Object> set = new LinkedTreeSet<>(Arrays.asList("A", "B"));

    set.removeAll(Arrays.asList("B", "C", "D"));

    assertThat(set).containsExactly("A");
    assertThat(set).hasSize(1);
}

@Test
public void testSizeAfterRetainAll() throws Exception {
    Set<Object> set = new LinkedTreeSet<>(Arrays.asList("A", "B"));

    set.retainAll(Arrays.asList("B"));

    assertThat(set).containsExactly("B");
    assertThat(set).hasSize(1);
}
```

**Comment by bwaldvogel** [comment]
Thanks, I’ll look into this…

**Comment by bwaldvogel** [comment]
Indeed very nasty bug. It should be fixed in 91ab3ec.

**Comment by markbigler** [comment]
I can confirm that [91ab3ec](https://github.com/bwaldvogel/mongo-java-server/commit/91ab3eca1cfa2f0c22d9faacf8b2271b556513e7) fixes the problem.

**Comment by bwaldvogel** [comment]
I’ve just released version `1.23.0` that contains this fix.
